sound/soc/sof/intel/hda.c

```diff
@@ -1057,8 +1103,27 @@ static int hda_sdw_machine_select(struct snd_sof_dev *sdev)
 	 */
 	if (link_mask && !pdata->machine) {
 		mach = pdata->desc->alt_machines;
-		while (mach && mach->link_mask && mach->link_mask != link_mask)
-			mach++;
+		while (mach && mach->link_mask) {
+			if (mach->link_mask != link_mask) {
+				mach++;
+				continue;
+			}
+			/* No need to match adr if there is no links defined */
+			if (!mach->links)
+				break;
```

Comment:
this would be an error, no?

Reply:
> this would be an error, no?

In my opinion, mach->links is not mandatory. links is only needed if there are two different machines with the same link mask, right? We can assume a default machine driver without links defined.

Reply:
that'd be fine, but we need to document that the use of the 'links' is required when 'link_mask' is not enough, e.g. when it's not accurate (Olympic) or when there are more than one device per link, or when the types of devices vary given the same link mask.